frontend/src/app/services/event-list.ts

```diff
@@ -4,8 +4,43 @@ import {dateDiff} from './utils';
 import {formatNumber} from '../utils/format';
 
 function seasonalDateDiff(date: string, mmdd: string) {
-  const date2 = date.slice(0, 5) + mmdd;
-  return dateDiff(date, date2);
+  return dateDiff(date, dateWithYear(mmdd, date));
+}
+
+/**
+ *
+ * @param date - date to append the year to (either MM-DD or YYYY-MM-DD)
+ * @param authDate - date from which to take the year
+ */
+function dateWithYear(date: string, authDate: string) {
+  if (date.length === 5) {
+    return authDate.slice(0, 5) + date;
+  } else if (date.length === 10) {
+    return date;
+  } else {
+    throw Error('Bad date format (should be YYYY-MM-DD or MM-DD)');
+  }
+}
+
+/**
+ * Generate first true randomly between given dates
+ * TODO: unit test
+ * @param date - simulation date
+ * @param dateFrom - interval start (either MM-DD or YYYY-MM-DD)
+ * @param dateTo - interval end (either MM-DD or YYYY-MM-DD)
+ */
+function dateBetweenTrigger(date: string, dateFrom: string, dateTo: string) {
+  if (seasonalDateDiff(date, dateFrom) < 0) {
+    // before given interval
+    return false;
+  } else if (seasonalDateDiff(date, dateTo) < 0) {
+    // in between interval
+    // const intervalDays = 1 + dateDiff(dateWithYear(dateTo, date), dateWithYear(dateFrom, date));
+    return (1 / (-seasonalDateDiff(date, dateTo))) > Math.random();
```

Comment:
What is the logic behind this? We are testing x∈[1/365, 1] for being greater than `rand([0, 1))`.

Reply:
This picks random date uniformly between now (date) and dateTo. (though I think we should include dateTo in the interval and change this to
```suggestion
    return (1 / (1 - seasonalDateDiff(date, dateTo))) > Math.random();
```

Reply:
It is better for 1 day before dayTo.

Yet, the concept of the function is hard to understand. In fact, it does: never before dateFrom, always after dateTo, and if between, the probability is inversely proportional to the proportion of the year we are before dateFrom. And it's called `dateBetweenTrigger`. _Between_. Nobody will understand it tomorrow, even if he does today.

Additionally, the whole thing is extremely error-prone, just try `dateBetweenTrigger('2020-01-01', '12-15', '01-15') // false`.

So if you wanted a recommendation, define, what you expect in the event definitions first, and then let's design the minimalistic function set for the purpose.

Reply:
We want a function that would fire _exactly_ once, uniformly distributed over given interval of days (represented by `dateFrom` and `dateTo`).
(imagine 0 0 0 0 1 0 0 0 as a time series output)
Unfortunately this seems impossible without providing context. (It would be an easy task if random number was generated on the start of interval and every day we would be able to access that number)

As we can use the `timeout` to regulate re-occurances of the event, the specification may be modified to:
We want a function that would fire for first time uniformly distributed over given interval of days.
(imagine 0 0 0 0 1 0 1 1 as a time series output)
Actually the 1/(1-daysRemaing) isn't proper probability of one day in the time series for the probability to be uniform but it assures very important aspect: The function must fire in the given interval at least once!

I'm using the second definition (first time fire matters), that's why I put 1 after the interval ends.
I agree this is confusing but it cannot be improved much without providing the context (generating random number/seed before). Maybe @daeglin might help with this?

Reply:
It was me who told Adam to implement it like this. It is a very common interview question in "quant world" to find online algorithm (in this sense https://en.m.wikipedia.org/wiki/Online_algorithm) for uniform sampling. So I didn't consider it weird at all but now I am realizing I am biased. I think a proper comment will solve it.

Reply:
Online algorithm for uniform sampling seems clear to me. Thanks for mentioning it. So the important part is, that we cut the tests after 1st success using timeout, right?

What I still do not understand is, why you do return `true` after dayTo. Is that important?